mistralrs-core/src/utils/model_config.rs

```diff
@@ -0,0 +1,328 @@
+        let mut xlora_config: Option<XLoraConfig> = None;
+        if is_xlora {
+            xlora_paths = vec![paths.get_classifier_path().as_ref().unwrap().to_path_buf()];
+            xlora_config = Some(paths.get_classifier_config().as_ref().unwrap().clone());
+        }
+
+        // Create VarBuilder:
+        // TODO: `from_mmaped_safetensors` has `xlora_paths` as the 2nd param, is this a bug?
+        let vb = from_mmaped_safetensors(
+            xlora_paths,
+            paths
+                .get_adapter_filenames()
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|(_, x)| (*x).to_owned())
+                .collect::<Vec<_>>(),
+            candle_core::DType::F32,
+            device,
+            silent,
+        )?;
+
+        Ok(Self {
+            lora_config,
+            xlora_config,
+            vb,
+            ordering,
+            preload_adapters,
+        })
+    }
+}
+
+// New type wrappers that segment the distinct parameter sets used by `from_ggml()` + `from_gguf()` methods:
+pub struct GGML(pub FileGGML);
+pub struct GGUF<'a>(pub FileGGUF<'a>, pub Device<'a>);
+
+// A `None` type vs the `Some` type (`Adapter<'a>`)
+pub struct NoAdapter {}
+
+// Marker traits to restrict type input:
+// (required workaround to support impl on subtypes, otherwise would use an enum)
+pub trait QuantParams {}
+impl QuantParams for GGML {}
+impl QuantParams for GGUF<'_> {}
+
+// Emulates `Option<Adapter>` but is compatible as a type bound in `impl<T>` for Some vs None
+pub trait MaybeAdapter {}
+impl MaybeAdapter for Adapter<'_> {}
+impl MaybeAdapter for NoAdapter {}
+
+// `derive_more::From` provides a terser construction for enum variants of `ModelParams`.
+#[derive(derive_more::From)]
+pub struct Config<Q: QuantParams, A: MaybeAdapter> {
+    pub quant: Q,
+    pub adapter: A,
+}
+
+// NOTE: Variantly used for `.expect_quantized()` / `.expect_adapted()` methods
+// `where` clause required due to bug with inline bounds:
+// https://github.com/luker-os/variantly/pull/16
+#[derive(variantly::Variantly)]
+pub enum ModelParams<'a, Q> where Q: QuantParams {
+    Quantized(Config::<Q, NoAdapter>),
+    Adapted(Config::<Q, Adapter<'a>>),
+}
+
+// A `builder()` method is derived from the `new()` method and it's params (derived builder struct fields).
+// NOTE: Intended to be built via fluent API in a single line, cannot conditionally append params.
+// `.adapter(Adapter<' >)` or for conditional usage `.and_adapter(Option<Adapter<' >)` can be used.
+// Otherwise omitting an `.adapter()` call prior to calling `build()` is ok, defaults to `None`.
+#[buildstructor::buildstructor]
+impl<'a, Q: QuantParams> ModelParams<'a, Q> {
+    #[builder]
+    pub fn new<'b: 'a>(quant: Q, adapter: Option<Adapter<'b>>) -> Self {
+        match adapter {
+            None => Self::Quantized((quant, NoAdapter {}).into()),
+            Some(a) => Self::Adapted((quant, a).into()),
+        }
+    }
+}

```

Comment:
An example of `buildstructor`. Usage is simple as seen in the loaders, `builder()` + method per parameter. Once `build()` is called it'll run the logic in this method which produces the GGUF / GGML enum from earlier.

If you were ever to implement support for EXL2 or some other format, it should keep it fairly simple 👍